### PR TITLE
Break the debugger when a test failed

### PR DESCRIPTION
**Pull request by Naios**
* Currently only implemented for the MSVC debugger if attached

The main idea is that we instantly see which test failed if we execute a test inside a debugger.

I would like to hear your opinion about this:
* Maybe we could allow to enable/disable this from CMake.
* Maybe we could move this up to the test macros which could yield a better stack trace

### CODE REVIEW

**Comment by hkaiser** [comment]
I like this idea. However, I'd suggest to use the already existing facilities in HPX. See [here](https://github.com/STEllAR-GROUP/hpx/blob/master/src/exception.cpp#L371-L380) for an example usage. This code looks up whether there was a command line option `--hpx:attach-debugger=exception`. I think we could add a `--hpx::attach-debugger=test-failure` which would enable attaching a debugger for your use case. See here for the relevant spots:

- https://github.com/STEllAR-GROUP/hpx/blob/master/src/util/parse_command_line.cpp#L506-L511
- https://github.com/STEllAR-GROUP/hpx/blob/master/src/util/command_line_handling.cpp#L1057-L1078

So your code would look like:
```
if (hpx::get_config_entry("hpx.attach_debugger", "") == "test-failure")
{
    hpx::util::attach_debugger();
}
```

That would however require to extract the corresponding declaration into a separate header file: https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/util/command_line_handling.hpp#L93.

**Comment by Naios** [comment]
This sounds great. I thought it would be comfortable if this would work automaitically without requiring any argument, since many unit tests don't even parse their arguments.
But I also think that one should be able to disable this.

Do you think that there is a solution which solves both ideas?

**Comment by hkaiser** [comment]
@Naios Hmmm, I'd prefer if this is not `ON` by default. But I agree that many tests don't handle command line options, I didn't think of this...

Internally HPX relies on extracting the information about any possible `--hpx:attach-debugger=...` option from its internal configuration database (using the key `hpx.attach_debugger`). This allows to enable extracting this information from the execution environment as well. If we added 

    "attach_debugger = ${HPX_ATTACH_DEBUGGER}",

here: https://github.com/STEllAR-GROUP/hpx/blob/master/src/util/runtime_configuration.cpp#L197.

Now setting `HPX_ATTACH_DEBUGGER=test-failure` in the execution environment would trigger the desired functionality. Note that you could still over-write this by using a command line option if we added `off` as a valid argument for it (see [link](https://github.com/STEllAR-GROUP/hpx/blob/master/src/util/parse_command_line.cpp#L506-L511) posted earlier).

Similarly we could add a compile-time configuration option enabling this by default:
```
#if defined(HPX_HAVE_TESTS_ATTACH_DEBUGGER_ON_FAILURE)
    "attach_debugger = ${HPX_ATTACH_DEBUGGER:test-failure}",
#else
    "attach_debugger = ${HPX_ATTACH_DEBUGGER}",
#endif
```

**Comment by Naios** [comment]
@hkaiser Perfect, thanks for your quick reply. That was the implementation I thought about :+1:

**Comment by Naios** [comment]
@hkaiser I applied the requested changes:
* `attach_debugger()` was refactored out
* The option is now enabled by default if `HPX_WITH_ATTACH_DEBUGGER_ON_TEST_FAILURE` is set

File: hpx/util/lightweight_test.hpp [hunk 1]
```diff
@@ -59,31 +59,9 @@ struct fixture
...
-        {
-            case counter_sanity:
-                ++sanity_failures_; return;
-            case counter_test:
-                ++test_failures_; return;
-            default:
-                { HPX_ASSERT(false); return; }
-        }
-    }
+    void increment(counter_type c);
```
Comment (hkaiser) [inline hpx/util/lightweight_test.hpp]: Shouldn't this be `HPX_EXPORT`ed?

File: hpx/util/lightweight_test.hpp [hunk 2]
```diff
@@ -59,31 +59,9 @@ struct fixture
...
-        {
-            case counter_sanity:
-                return sanity_failures_;
-            case counter_test:
-                return test_failures_;
-            default:
-                { HPX_ASSERT(false); return 0; }
-        }
-    }
+    std::size_t get(counter_type c) const;
```
Comment (hkaiser) [inline hpx/util/lightweight_test.hpp]: Same here: `HPX_EXPORT`?

File: hpx/util/lightweight_test.hpp (hunk 1 above)
Comment (Naios) [inline hpx/util/lightweight_test.hpp]: I thought that too, however `lightweight_test.cpp` is built into the `hpx_init` static library and thus we don't need to export the symbol.

File: hpx/util/lightweight_test.hpp (hunk 2 above)
Comment (Naios) [inline hpx/util/lightweight_test.hpp]: Same as above

File: hpx/util/lightweight_test.hpp (hunk 1 above)
Comment (hkaiser) [inline hpx/util/lightweight_test.hpp]: Right, thanks!

**Comment by hkaiser** [comment]
@Naios: one more thing - may I ask you to add support for `--hpx:attach-debugger=off` (or similar) allowing to disable the pre-set `test-failure` if `-DHPX_WITH_ATTACH_DEBUGGER_ON_TEST_FAILURE=On`?

**Comment by Naios** [comment]
@hkaiser Yes, good idea. I'll add it.

**Comment by Naios** [comment]
@hkaiser I applied the requested change

File: src/util/command_line_handling.cpp [hunk 3]
```diff
@@ -1037,34 +1038,18 @@ namespace hpx { namespace util
...
         {
             std::string option = vm_["hpx:attach-debugger"].as<std::string>();
-            if (option != "startup" && option != "exception") {
+            if (option != "off" && option != "startup" &&
+                option != "exception" && option != "test-failure") {
                 std::cerr <<
                     "hpx::init: command line warning: --hpx:attach-debugger: "
                     "invalid option: " << option << ". Allowed values are "
-                    "'startup' or 'exception'" << std::endl;
+                    "'off, startup', 'exception' or 'test-failure'" << std::endl;
```
Comment (hkaiser) [inline src/util/command_line_handling.cpp]: Nitpick: this missed apostrophes

Comment (Naios) [reply src/util/command_line_handling.cpp]: /fixed

**Pull request merged**